teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,326 @@
+
+Example:
+
+```
+cluster://default/task/git-clone
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: "cluster"
+  name: "cluster://default/task/git-clone" 
+```
+
+##### Custom Resolver
+When implementing a new custom resolver, just like users have the freedom to choose their parameters, they also have the freedom to define their `name` (or `url`) based on:
+
+```
+<scheme>://<location>[@<version>][#<selector>][?<params>]
+```
+
+The custom resolver can choose the schemes it wants to subscribe to and for each scheme, it can define the remainder of `location, version, selectors and params` as needed. An example `scheme` could be `<custom-resolver-name>`. A `location` may be made up of concrete units e.g. `location = <foo>/<bar>` (i.e. expect two parts) that are up to the discretion of the custom resolvers.
+
+The custom resolver must validate the `name` which will be passed within the `Remote Resolution CRD's` spec. The `name` (or `url`) can then be parsed based on the format for that `scheme` and used to fetch the remote resource.
+
+### Stage 2
+
+In this stage, we enable the Tekton controller to be able to identify the resolvers on the clusters that have subscribed to the scheme of the incoming [resolver reference](#resolver-reference).
+
+#### Subscribing to `scheme`s
+
+Each remote resolver will subscribe to `scheme`s. e.g. the git resolver might subscribe to `scheme`s like `git+https`, `https` or a bundle resolver might subscribe to `scheme`s like `oci`.
+This mapping of `scheme` to `resolver` could be easily maintained in a `configMap` for the cluster. The cluster operators will be required to know the remote resolvers that are running on the cluster and the schemes they are subscribed to.  
+The configMap must list out all the schemes supported by all the resolvers in [order of priority](#prioritizing-the-order-of-resolvers).
+ 
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: remote-resolution-scheme-subscription
+  namespace: tekton-pipelines
+data:
+  #<scheme>: "<resolvers subscribed to the scheme; comma separated in order of priority>"
+  https: "git,http"
+  git+https: "git"
+  git: "git"
+  http: "http"
+  oci: "bundle"
+  artifacthub: "hub"
+```
+
+#### Prioritizing the order of resolvers
+
+Occasionally, for certain common `scheme`s like `https`, multiple resolvers could be subscribed to it. In that case, the Tekton controller would send out requests based on the priority to the list of the subscribed resolvers until one of them succedes in resolving the request. The cluster operators will maintain the priority of the resolvers based on the order of the resolvers subscribed to the scheme. In the comma-separated value, the left side has the higher priority. The advantage here is that depending on the `scheme`, the priority can be different and is under the control of the cluster operators.
```

Comment:
nit: the controller could send it to multiple but get the results in order, to optimize time consumption.

Reply:
Great point! It can federate the request to all of them.

This has triggered a thought:
Imagine, we didn't have to worry about a priority/scheme based subscription at all? 
What if we always federated any remote resolution request to all the resolvers? 
In this scenario, most resolvers that are not subscribed to the scheme would immediately fail. Only the ones that understand it would succeed. Is that bad? This would dramatically reduce the complexity.
Right now, the TaskRun/PipelineRun controller identifies which resolvers to forward the request to. Why not have it federate to them all? And then it can round robin and get the resource from the resolver that succeeds. Even if multiple resolvers succeed, there is no harm done, we have what we need.

Reply:
cc @wlynch WDYT?

Reply:
The downside of this would be, if it works as today, a growing number of objects (each "resolving" would mean a `ResolutionRequest` for each resolver). We tend to have etcd pressure already with the amount of objects we have so…

Reply:
That's a good point. Does the priority matter then? If we send out request to all the subscribed resolvers then as long as we get a response from any of them, that should be fine?

Reply:
I've updated this section. I don't think there is a need for a priority if we federate the request to all the subscribed resolvers simultaneously.

Reply:
It dramatically reduces the complexity since the cluster operators also don't need to maintain configMaps.

Reply:
PTAL at the updated section.